### DISCUSSION TITLE
Versioning for images to avoid confusion when mod_expires is in use

### DISCUSSION THREAD

**Discussion by madsoliver** [PrestaShop/PrestaShop#33406] [Ideas]
**The problem**
Some of my client pull their hair off trying to understand why for example category images are not updated in FrontOffice after they upload a new image. It’s off course caused by Apache mod_expires directives in the .htaccess.. That’s a good thing usually, but not when image names do not change after updates…. Accessing  `/img/c/5-category_default.jpg` once before an update, and then again after uploading a new image will give `304 Not Modified`, and the merchant does not understand why nothing has changed... However, I don’t want to turn off `mod_expires` since it’s good for page speed… 

**Possible solution:**
Couldn’t we add a versioning-system to images, like done with cached css and js?
I did some testing and, not surprisingly `/img/c/5-category_default.jpg?version=1234` or `/img/c/5-category_default.jpg?date_updated=849846548949` gives me `200 Ok` allthough `/img/c/5-category_default.jpg` was cached in the browser. Problem solved.
What do you think?

PS: There is probably a lot of more places than only category images that this is an issue.. So maybe a `date_updated` inside `ps_image`, and then just append that to all image links will do the trick for all images that use that table?

[a comment by an automated account is not shown]

**Comment by khouloudbelguith** [comment]
Hi @onklmaps,

Thanks for your report.
Ping @PrestaShop/product-team what do you think?

Thanks!

**Comment by rdy4ever** [comment]
Hi, @onklmaps 
Actually, query string versioning is not recommended, as it creates browser caching issues. (and we do need images to be cached by the browser for performance). We could add versioning in the file name like `/img/c/5-category_default-v[insert UNIX timestamp here].jpg` This is the recommended way for all file versioning. Indeed, it will create redundant files, so we also need a way to delete the old file, to spare storage space.

**Comment by madsoliver** [comment]
Hi @rdy4ever 
Maybe we can keep one file only (and replace on update), but rename the file only in the url - then mod_rewrite to original file?
I might be in deep water here.

**Comment by ghost** [comment]
Hi, 

It's a nice improvement, I also have this problem, favicon also use a timestamp on the classic theme.

With which browser there is an issue ?

You can do this easy with the date_upd field 

`<img src="{$category.image.large.url}?v={$category.date_upd|date_format:'%m%d%y%M'}" alt="{if !empty($category.image.legend)}{$category.image.legend}{else}{$category.name}{/if}">`

**Comment by PierreRambaud** [comment]
I agree the current system is not really appropriate.

Afaik, the image name is not important (not anymore), we could create apache rewrite rule in order to integrate the date_upd inside the path, wdyt?

**Comment by ghost** [comment]
Yes it's a solution, but images rewrite are locked by https://github.com/PrestaShop/PrestaShop/pull/21196

**Comment by PierreRambaud** [comment]
Nice @okom3pom forget this one, don't forget to help the QA team to provide a way to generate a lot of images (ofc if you can) :sweat_smile:

**Comment by hibatallahAouadni** [comment]
Hello @onklmaps 

It's been more than **26 days** and you didn't answer @PierreRambaud :confused: 
Usually, I close the issue with no response more than **20 days**. So, **please answer ASAP**.
@okom3pom and @PierreRambaud if you think we should to add the label **`To Do`** even without the author confirmation, please go ahead :wink: 

Thanks!

**Comment by PierreRambaud** [comment]
It's still in To Be Specified, we don't have the right way to do it, so it can't be in To Do :thinking:

**Comment by madsoliver** [comment]
> I agree the current system is not really appropriate.
> 
> Afaik, the image name is not important (not anymore), we could create apache rewrite rule in order to integrate the date_upd inside the path, wdyt?

Sorry for the delay.
Yes, I think this is a good solution 🙂

**Comment by ghost** [comment]
> Yes it's a solution, but images rewrite are locked by #21196

Oh I said something stupid, it's the products rewrite rules who are stuck not category

**Comment by zorrobiwan** [comment]
BTW, this issue is also there when using Media Servers (like CloudFront).
When changing Category image, you **_need_** to invalidate the CDN manually (or wait) to get the new image which is sometimes **not possible** for the BO operator.

So, waiting for a good solution too :)
Thx
